File: benchmark/julia/src/langevin.jl
using Random
using ThreadsX

struct Langevin{F,G}
    x₀::Float64
    f::F
    g::G
    function Langevin(f::F, g::G; x₀::Float64=0.0) where {F,G}
        new{F,G}(x₀, f, g)
    end
end

function simulate(eq::Langevin, T::Float64, τ::Float64=0.01)
    t = collect(0:τ:T)
    n = length(t) - 1
    x = zeros(n + 1)
    x[1] = eq.x₀
    @inbounds for i in 1:n
        dw = eq.g(x[i], t[i+1]) * sqrt(τ) * randn()
        x[i+1] = x[i] + eq.f(x[i], t[i+1]) * τ + dw
    end
    t, x
end

function displacement(eq::Langevin, T::Float64, τ::Float64=0.01)
    n = ceil(Int, T/τ)
    current_x = eq.x₀
    current_t = 0.0
    for _ in 1:n-1
        dw = eq.g(current_x, current_t) * sqrt(τ) * randn()
        current_x += eq.f(current_x, current_t) * τ + dw
        current_t += τ
    end
    last_step = T - current_t
    dw = eq.g(current_x, current_t) * sqrt(last_step) * randn()
    current_x += eq.f(current_x, current_t) * last_step + dw
    current_x - eq.x₀
end


function msd(eq::Langevin, T::Float64, N::Int=10_000, τ::Float64=0.01)::Float64
    displacements = ThreadsX.map(1:N) do _
        displacement(eq, T, τ) ^ 2
    end
    return sum(displacements) / N
end

eq = Langevin((x, t) -> -x, (x, t) -> 1.0)
T = collect(100.0:100.0:1000.0)
@time m = [msd(eq, t) for t in T]
println(m)

File: benchmark/python/langevin.py
from math import sqrt
from typing import Callable

import numpy as np
from numba import (  # pyright: ignore [reportMissingTypeStubs]
    njit,  # pyright: ignore [reportUnknownVariableType]
    prange,
)


class Langevin:
    def __init__(
        self,
        f: Callable[[float, float], float],
        g: Callable[[float, float], float],
        starting_point: float = 0.0,
    ):
        self.starting_point: float = starting_point
        self.drift: Callable[[float, float], float] = f
        self.diffusivity: Callable[[float, float], float] = g

    def simulate(self, duration: float, tau: float = 0.01):
        return _simulate(
            self.drift,
            self.diffusivity,
            duration,
            self.starting_point,
            tau,
        )

    def msd(
        self,
        duration: float,
        N: int = 10_000,
        tau: float = 0.01,
    ):
        return _msd(
            self.drift,
            self.diffusivity,
            duration,
            self.starting_point,
            N,
            tau,
        )


@njit
def _simulate(
    f: Callable[[float, float], float],
    g: Callable[[float, float], float],
    duration: float,
    starting_point: float = 0.0,
    tau: float = 0.01,
):
    t = np.arange(0, duration, tau)
    n = len(t) - 1
    noise = np.random.normal(0, 1, n)
    x = np.zeros((n + 1,))
    x[0] = starting_point
    for i in range(1, n + 1):
        drift_term = f(t[i - 1], x[i - 1]) * tau  # pyright: ignore [reportAny]
        diffusion_term = g(t[i - 1], x[i - 1]) * noise[i - 1] * sqrt(tau)  # pyright: ignore [reportAny]
        x[i] = x[i - 1] + drift_term + diffusion_term
    x = np.cumsum(noise)
    return t, x


@njit(parallel=True)  # pyright: ignore [reportUntypedFunctionDecorator]
def _msd(
    f: Callable[[float, float], float],
    g: Callable[[float, float], float],
    duration: float,
    starting_point: float = 0.0,
    N: int = 10_000,
    tau: float = 0.01,
):
    total = 0.0
    for _k in prange(N):
        _, x = _simulate(f, g, duration, starting_point, tau)
        total += (x[-1] - x[0]) ** 2  # pyright: ignore [reportAny]
    return total / N


if __name__ == "__main__":
    import time

    @njit
    def f(_t: float, x: float) -> float:
        return -x

    @njit
    def g(_t: float, _x: float) -> float:
        return 1.0

    duration = [i for i in range(100, 1100, 100)]
    eq = Langevin(f, g)
    _ = eq.msd(1.0)
    msds = [0.0 for _ in duration]
    start = time.time()
    for i, d in enumerate(duration):
        msds[i] = eq.msd(d)
    elapsed_parallel = time.time() - start
    print(f" Time elapsed: {elapsed_parallel:.3f} seconds")
